### PR TITLE
Start using Rope for attribute run storage

### PR DESCRIPTION
**Pull request by lorentey**
- Replace `Array<_InternalRun>` with `Rope<_InternalRun>`.
- Fix semantics of AttributeRunBoundaries.character. It no longer pretends that attributes can be tied to grapheme clusters — that never actually worked properly, and it cannot be supported without breaking the intended use case.
- Remove cache of latest run position; log(n) might be fast enough that we don’t need to worry about that.
- Switch AttributedString.Guts members to take/return BigString.Index values, not AttributedString indices. This cuts down on constant back-and-forth conversions.
- Discard unused Guts members; update remaining mutation methods to follow Swift naming conventions.
- Stop using the `public extension` language misfeature.

rdar://108152217

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

File: Sources/FoundationEssentials/AttributedString/AttributedString+_InternalRuns.swift [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        return (Index(r.index, offset: offset, utf8Offset: utf8Offset - r.remaining), r.remaining)
+    }
+
+    func _exactIndex(atUTF8Offset utf8Offset: Int) -> Index {
+        let r = index(atUTF8Offset: utf8Offset)
+        precondition(r.remaining == 0)
+        return r.index
+    }
+
+    func index(containing i: AttributedString.Index) -> (index: Index, utf8Offset: Int) {
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedString+_InternalRuns.swift:214]: Rename `utf8Offset` to `remaining` or `utf8Remaining`.

File: Sources/FoundationEssentials/AttributedString/AttributedString+_InternalRunsSlice.swift [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+    var isEmpty: Bool {
+        _utf8Bounds.isEmpty
+    }
+
+    var startIndex: Index {
+        _guts.runs.index(atUTF8Offset: _utf8Bounds.lowerBound).index
+    }
+
+    var endIndex: Index {
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedString+_InternalRunsSlice.swift:59]: Document this

File: Sources/FoundationEssentials/AttributedString/AttributedString.swift [hunk 3]
```diff
@@ -316,17 +326,20 @@ extension AttributedString {
...
-            var substr = AttributedSubstring(_guts, in: bounds.relative(to: unicodeScalars))
+            let bounds = bounds.relative(to: unicodeScalars)
+            var substr = AttributedSubstring(_guts, in: bounds._bstringRange)
             let ident = Self._nextModifyIdentity
             substr._identity = ident
-            _guts = Guts(string: "", runs: []) // Dummy guts so the substr has (hopefully) the sole reference
+            _guts = Guts() // Dummy guts to allow in-place mutations
             defer {
                 if substr._identity != ident {
+                    // FIXME: Why is this necessary?
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedString.swift:342]: This just prevents affecting data outside the range

File: Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift [hunk 4]
```diff
@@ -82,124 +82,147 @@ extension AttributedString.Guts {
...
+        return i ..< block.end!
     }
     
     // MARK: Attribute Utilities
     
     private func _constrainedAttributes(
-        at location: Int, with constraint: AttributeRunBoundaries
+        at utf8Offset: Int, with constraint: AttributeRunBoundaries
     ) -> _AttributeStorage {
         // Don't update the cache, because getting constrained attributes might look backwards very far and we'll just have to iterate the cache back to where we currently are
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift:110]: Remove this comment

File: Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift [hunk 5]
```diff
@@ -82,124 +82,147 @@ extension AttributedString.Guts {
...
+        type: AttributedString.AttributeRunBoundaries,
+        from utf8Offset: Int,
+        to utf8Range: Range<Int>
     ) {
-        let style = _constrainedAttributes(at: idx, with: type)
-        enumerateRuns(containing: range) { run, _, _, _ in
-            run.attributes.matchStyle(of: style, for: type)
+        let style = _constrainedAttributes(at: utf8Offset, with: type)
+        runs(in: utf8Range).updateEach { attributes, _, _ in
+            attributes.matchStyle(of: style, for: type)
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift:154]: TODO: See if it makes sense for this to set `modified` to false sometimes

File: Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift [hunk 6]
```diff
@@ -217,32 +240,36 @@ extension AttributedString.Guts {
...
         }
 
         // Paragraph-based constraints
         if type == .attributes && constraintsInvolved?.contains(.paragraph) ?? true {
             // Attributes are always applied consistently, so we only need to expand outwards and not fix the range of the mutation itself
-            let paragraphStyle = _constrainedAttributes(at: range.lowerBound, with: .paragraph)
+            let paragraphStyle = _constrainedAttributes(at: utf8Range.lowerBound, with: .paragraph)
             let paragraphRange = _paragraph(in: strRange)._utf8OffsetRange
             // FIXME: It looks like this assumes that mutated attributes are consistent throughout
             // FIXME: the mutated range. This expectation should be explicitly documented.
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift:263]: This FIXME can be resolved -- this method is only called when the mutation is affecting attributes only. Replace with a comment highlighting this.

File: Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift [hunk 7]
```diff
@@ -217,32 +240,36 @@ extension AttributedString.Guts {
...
-            enumerateRuns(containing: paragraphRange.lowerBound ..< range.lowerBound) { run, _, _, _ in
-                run.attributes.matchStyle(of: paragraphStyle, for: .paragraph)
+            self.runs(
+                in: paragraphRange.lowerBound ..< utf8Range.lowerBound
+            ).updateEach { attributes, _, _ in
+                attributes.matchStyle(of: paragraphStyle, for: .paragraph)
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift:267]: TODO: See if it makes sense for this to set modified to false sometimes

File: Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift [hunk 8]
```diff
@@ -217,32 +240,36 @@ extension AttributedString.Guts {
...
             }
-            enumerateRuns(containing: range.upperBound ..< paragraphRange.upperBound) { run, _, _, _ in
-                run.attributes.matchStyle(of: paragraphStyle, for: .paragraph)
+            self.runs(
+                in: utf8Range.upperBound ..< paragraphRange.upperBound
+            ).updateEach { attributes, _, _ in
+                attributes.matchStyle(of: paragraphStyle, for: .paragraph)
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift:272]: TODO: See if it makes sense for this to set modified to false sometimes

File: Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift [hunk 9]
```diff
@@ -308,102 +335,47 @@ extension AttributedString.Guts {
...
+        while i < string.endIndex {
             let j = nextParagraphBreak(after: i)
-            let startOffset = utf8Offset(of: i)
-            let endOffset = utf8Offset(of: j)
-            let paragraphStyle = self._constrainedAttributes(at: startOffset, with: .paragraph)
-            self.enumerateRuns(containing: startOffset ..< endOffset) { run, _, _, mod in
-                run.attributes.matchStyle(of: paragraphStyle, for: .paragraph)
+            let paragraphStyle = self._constrainedAttributes(at: i.utf8Offset, with: .paragraph)
+            self.runs(in: i.utf8Offset ..< j.utf8Offset).updateEach { attributes, _, _ in
+                attributes.matchStyle(of: paragraphStyle , for: .paragraph)
```
Comment (lorentey) [inline Sources/FoundationEssentials/AttributedString/AttributedStringAttributeConstrainingBehavior.swift:378]: TODO ditto

**Comment by lorentey** [comment]
@swift-ci test

**Review by jmschonfeld — APPROVED** [review]

**Comment by lorentey** [comment]
Mac failures are unrelated -- merging based on Linux results.

**Pull request merged**